PathPayment

bartekn opened the pull request:
This PR adds PathPayment operation mutators. Check [`ExamplePathPayment` function](https://github.com/stellar/go-stellar-base/pull/20/files#diff-b0bd7cdedee15ec71276b9561bda7cbc).

nullstyle commented:
I'm going to look over this some more, but on its head I don't think this is the correct API that we should provide.

The point of the mutators is to reduce the number of nouns that an integrator has to keep in their head and to promote code reuse.  In addition, the build api should aim to provide sensible defaults.

As an example...  having `PathAmount` is doesn't really need to exist... instead `NativeAmount` and `CreditAmount` should both implement `PathPaymentMutator`.

I'll make some specific comments on the code itself, but I just wanted to give you an introduction to what I think we should change.

nullstyle commented on build/main.go:
```diff
@@ -92,6 +99,22 @@ type NativeAmount struct {
+}
+
+// PathDestination is a mutator that configures a path_payment's destination asset and amount
+type PathDestination struct {
```
As mentioned, this doesn't need to exist.  Someone should simply use the normal `Destination` mutator along with one of `NativeAmount` or `CreditAmount` to replace the functionality of this mutator.

bartekn replied:
That was my first idea but the PathPayment contains SendMax amount too.

nullstyle replied:
I'm only referring to the `PathDestination` struct in the comment above... It doesn't deal source portion of the payment.

So, I tried to design this package so that the onramp is gradual...  Lessons learned crafting one kind of transaction can be used when crafting other transactions.  IMO, that suggests that rather than replacing `Destination`, we introduce a modifier that augments the built transaction.  Here's the flow I see:

1.  If a payment has not destination information (asset/amount) an error occurs
2.  If a payment only has destination information, a `PaymentOp` is created when applied
3.  If a payment has source information, instead a `PathPaymentOp` will be created when applied, with a default empty path
4.  If a payment has path information, use it instead of the default.

By doing this, my hunch is that developers will be able to gradually learn the api, eventually learning how to craft more advanced transactions by adding new mutators to their repertoire instead of having to learn separate "constellations" of structs that only interact with one another to perform a given advanced transaction.

Doe that make sense?

bartekn replied:
> I'm only referring to the `PathDestination` struct in the comment above... It doesn't deal source portion of the payment.

There's `PathSend` struct below. It all makes sense, I was saying that with the current design it wasn't possible to set both SendMax and DestAmount the API I created was one of possible workarounds. I like your new design but I'd change one thing. Check my previous comment added to your snippet.

nullstyle commented on build/main.go:
```diff
@@ -92,6 +99,22 @@ type NativeAmount struct {
 	Amount string
 }
 
+type Path struct {
```
IMO, the build `package` should be about ease-of-use.  IMO, building up a big nested object graph using go literals is a little heavy, especially in the case where there is small path involved:

```
Path{ Assets: []Asset{ Asset{Native: true} } }
```

As opposed to one of the following:

```
Path().Native()
Path{}.Native() // I think this is valid provided we use non-pointer method definitions
```

The equivalent of your example send would be:
```
Path().Native().Credit("BTC", "GAHJZHVKFLATAATJH46C7OK2ZOVRD47GZBGQ7P6OCVF6RJDCEG5JMQBQ")
```

nullstyle commented on build/main_test.go:
```diff
@@ -34,3 +34,45 @@ func ExampleTransactionBuilder() {
+
+// ExamplePathPayment creates and signs a simple transaction with PathPayment operation, and then
+// encodes it into a base64 string capable of being submitted to stellar-core.
+func ExamplePathPayment() {
```
Let me give an alternative formulation of this example and then I'll comment on some of the possible changes I introduce:

```go

nullstyle replied:
Notice that I reuse the normal "Amount" mutators to provide the source asset and the source max.  It's a bit awkward with regards to communicating the fact that the source amount is a maximum, but IMO the consistency of use is important.  If we were to go your route using the `Asset` as a separate struct, IMO we should do the same thing with the non-path payment.  That is to say `NativeAmount` and `CreditAmount` would be replaced with 3 mutators:  `NativeAsset{}`, `CreditAsset{Code,Issuer}` and `Amount{string}`.

nullstyle replied:
Notice that the Path builder object can have chained calls to build the path.  One alternative formulation, provided we went with the `NativeAsset` change proposed in the comment above:  

```
Path(NativeAsset{}, CreditAsset{"USD", "Blah"})
```

nullstyle replied:
Notice that there is no separate `PathPayment` mutator.  Instead, I use the normal `Payment` mutator and it knows that when I start provided path payment specific mutators that it should build a PathPaymentOp instead.  This allows the end developer to not care whether to pick one or the other.  The system will just use the most appropriate operation.

IMO, this is consistent with how we expose non-native payments in the higher level SDKs...  We paint over some of the specifics to give a simpler external interface.  For example, js-sdk's `Asset` doesn't differentiate between alphanum4 and alphanum12 assets... it will just transparently use the correct one based upon the code supplied.

bartekn replied:
I agree with all of your points but I don't like the way we're going to define [send, dest] x [assets, amounts]. First, I think that with this API it's quite easy to mistake send and dest amounts ("was it `Source()` or `Dest()` function?"). Second, as you said, it looks a bit awkward.

What if `Path` mutator was a struct like this (similar thing to what @irisli did in Laboratory):

```go 
type Path struct {
	SendMax string
	DestinationAmount string
	Assets []{}interface // This will always have at least 2 elements: send and dest asset
}

type (p Path) setAssets(assets ...interface{}) Path {
   p.Assets = assets
   return p
}
```

Then, when it comes to creating a path we could do something like this:
```go
Payment(
	Destination{"GBDT3K42LOPSHNAEHEJ6AVPADIJ4MAR64QEKKW2LQPBSKLYD22KUEH4P"},
	Path{
		SendMax: "10",
		DestAmount: "20",
	}.setAssets(NativeAsset{}, CreditAsset{"USD", "Blah"}),
)
```
Passing `Path` mutator to a PaymentBuilder will switch to underlying `PathPaymentOp`.

nullstyle replied:
Personally, I don't like it.  Conceptually, it replaces the existing `Amount` mutators (in a path payment context) which is one thing I was trying to avoid.  Think "Small pieces working together", something like the unix philosophy.

IMO, a payment within the stellar network is defined by three sets of information:  The destination information, the source information, and the intermediary path information.  I think that by having these three as separate mutators/concepts gives us the most flexibility and will let more feature-rich mutators be built through composition.

---

Also, one quick aside.  I've tried to replace "SendMax" in its various forms with  "SourceMax", IMO it consistent with the notion that every payment has the mentioned three sets of info.

I found myself numerous times getting confused and misusing "source" when it should be "send", so I feel it's a good change to make higher level APIs more understandable.  Note:  The documentation should still mention it sets the "SendMax" field when talking about how the underlying xdr operation is mutated.

---

If you believe that people will confuse "source" and "destination", I think we should work to disambiguate rather than abandon the concept of the gradual onramp.  I've got a more drastic change that I'll write up an post momentarily, but let me also offer some augmentations of my proposal above.    Instead of `Source()` how about `PayWith` or `PayUsing`?:

```go
Payment(
  Destination{"GBDT3K42LOPSHNAEHEJ6AVPADIJ4MAR64QEKKW2LQPBSKLYD22KUEH4P"},
  CreditAmount{ "EUR", "GCPZJ3MJQ3GUGJSBL6R3MLYZS6FKVHG67BPAINMXL3NWNXR5S6XG657P", "100" },
  PayWith(CreditAmount{ "USD", "GAWSI2JO2CF36Z43UGMUJCDQ2IMR5B3P5TMS7XM7NUTU3JHG3YJUDQXA", "50" }),
  Path().Native().Credit("BTC", "GAHJZHVKFLATAATJH46C7OK2ZOVRD47GZBGQ7P6OCVF6RJDCEG5JMQBQ"),
)
``` 

Regarding the path: The "gradual onramp" goal is one reasons that I am consciously keeping "path" mutators separate from "source".  It allows someone to learn that "A payment can be made from one asset to another" without having to understand that the system can hop through multiple assets as well.

nullstyle replied:
My more drastic proposal to resolve the source/dest ambiguity:

```go
SendNative{Amount: "100.0"}
SendCredit{Code:"USD", Issuer:"BLAH" Amount:"100.0"}
PayWithNative{MaxAmount: "100.0"}
PayWithCredit{Code:"USD", Issuer:"BLAH", MaxAmount: "100.0"}
```

And then one additional proposal about how we might expose pathing:

```
// manually specified path
Payment(
  Destination{addy},
  PayWithNative{MaxAmount: "100.0"}.
    ThroughCredit("EUR", "ISSUER").
    ThroughCredit("USD", "SOMEONEELSE"),  
)

// Path finding
Payment(
  horizon.FindPath(
    SendNative{Amount: "100.0"}
    PayWithCredit{Code:"USD", Issuer:"BLAH"} // MaxAmount will be filled automatically if a path is found using USD this source asset
  ),
)
```

Thoughts?

nullstyle commented:
So, I wanted to close with a vision of where we keep the API as modular and adaptable as I hope:

```go
// A vision of when the ecosystem is more fleshed out.
func ExamplePathPayment() {
	seed := "SDOTALIMPAM2IV65IOZA7KZL7XWZI5BODFXTRVLIHLQZQCKK57PH5F3H"
  server := horizon.NewClient("https://stellar.org")
  
	tx := Transaction(
    SourceAccount{seed},
		server.LatestSequence(seed), // this automatically looks up the latest sequence for the account
		Payment(
      // this automatically performs a path finding operation and uses the most economic path.
      // IMO, we probably need to add some way to pre-approve what source asset to use and error in a sitation
      // where we cannot decide automatically what source asset to use.
      server.BestPath(
        Destination{"GBDT3K42LOPSHNAEHEJ6AVPADIJ4MAR64QEKKW2LQPBSKLYD22KUEH4P"},
        CreditAmount{ "EUR", "GCPZJ3MJQ3GUGJSBL6R3MLYZS6FKVHG67BPAINMXL3NWNXR5S6XG657P", "100" },
        SourceAccount{seed},
      )
		),
	)

	txe := tx.Sign(seed)
	txeB64, _ := txe.Base64()

	fmt.Printf("tx base64: %s", txeB64)
	// Output: tx base64: AAAAADZY/nWY0gx6beMpf4S8Ur0qHsjA8fbFtBzBx1cbQzHwAAAAZAAAAAAAAAABAAAAAAAAAAAAAAABAAAAAAAAAAIAAAABVVNEAAAAAAAtJGku0Iu/Z5uhmUSIcNIZHodv7Nkv3Z9tJ02k5t4TQQAAAAAdzWUAAAAAAEc9q5pbnyO0BDkT4FXgGhPGAj7kCKVbS4PDJS8D1pVCAAAAAUVVUgAAAAAAn5TtiYbNQyZBX6O2LxmXiqqc3vheBDWXXttm3j2Xrm8AAAAAO5rKAAAAAAIAAAAAAAAAAUJUQwAAAAAADpyeqirBMAJpPzwvuVrLqxHz5shND7/OFUvopGIhupYAAAAAAAAAARtDMfAAAABAvbv3jz4I/xVsIxamvzyOaa048aJZHmpLpjny5b0pe3ln1geH593O5jOoo4CPBi2AsgqlnAO55cz8yMQ0ibuCAQ==
}
```

nullstyle commented:
As a further example of higher-level mutators we might expose:

I would love to see a sibling to the `SourceAccount` mutator that loads the secret key from vault (https://www.vaultproject.io/), or perhaps a locally encrypted wallet file.

bartekn commented:
I updated this PR and the current API is the following:
```go
tx := Transaction(
		SourceAccount{seed},
		Sequence{1},
		Payment(
			Destination{"GBDT3K42LOPSHNAEHEJ6AVPADIJ4MAR64QEKKW2LQPBSKLYD22KUEH4P"},
			CreditAmount{"USD", "GAWSI2JO2CF36Z43UGMUJCDQ2IMR5B3P5TMS7XM7NUTU3JHG3YJUDQXA", "50"},
			PayWith(CreditAsset("EUR", "GCPZJ3MJQ3GUGJSBL6R3MLYZS6FKVHG67BPAINMXL3NWNXR5S6XG657P"), "100").
				Through(NativeAsset()).
				Through(CreditAsset("BTC", "GAHJZHVKFLATAATJH46C7OK2ZOVRD47GZBGQ7P6OCVF6RJDCEG5JMQBQ")),
		),
	)
```

* I implemented `DestAsset` and `DestAmount` as you suggested,
* I decided to leave `Asset` struct because we will need this in `ChangeTrust` and `ManageOffer` operations. I added two helper methods that simplify creating `Asset` objects:
  ```go
  // Helper method to create native Asset object
  func NativeAsset() Asset {
	  return Asset{Native: true}
  }
  
  // Helper method to create credit Asset object
  func CreditAsset(code, issuer string) Asset {
	  return Asset{code, issuer, false}
  }
  ```
* When it comes to setting `SendMax` and `SendAsset` I had to create `PayWith(sendAsset Asset, maxAmount string)` helper method because it's not possible to do something like this in go:
  ```go
  PayWith{"USD", "BLAH", "50"}.Through(NativeAsset())
  ```

  This method simply returns `PayWithPath` struct with populated fields. `PayWithPath` has `Through` method that can be used to specify path.

Let me know what you think.

nullstyle commented:
That looks really great to me.  There might be some confusion about when to use `CreditAmount` vs. `CreditAsset`, but that seems small to me.  

+1

bartekn commented:
> There might be some confusion about when to use `CreditAmount` vs. `CreditAsset`

The reason for using `CreditAsset` instead of `CreditAmount` in `PayWith` is to stress that the amount sent in this transaction won't be exactly the same as amount set (thus the `MaxAmount` field name in `PayWith`).
